pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -187,40 +212,126 @@ func (c *AntreaIPAMController) cleanIPPoolForStatefulSet(namespacedName string)
+	if exists {
+		// Stateful Set Pod is annotated with dedicated IP pool
+		return strings.Split(annotations, annotation.AntreaIPAMAnnotationDelimiter)
+	}
+
+	// Inspect Namespace
+	namespace, err := c.namespaceLister.Get(ss.Namespace)
+	if err != nil {
+		// Should never happen
+		klog.Errorf("Namespace %s not found for StatefulSet %s", ss.Namespace, ss.Name)
+		return nil
+	}
+
+	annotations, exists = namespace.Annotations[annotation.AntreaIPAMAnnotationKey]
+	if exists {
+		return strings.Split(annotations, annotation.AntreaIPAMAnnotationDelimiter)
+	}
+
+	return nil
+
+}
+
+// Look for an IP Pool associated with this StatefulSet, either a dedicated one or
+// annotated to the Namespace. If such IP Pool is found, preallocate IPs for the StatefulSet.
+// This function returns error if pool is not found, or allocation fails.
+func (c *AntreaIPAMController) preallocateIPPoolForStatefulSet(ss *appsv1.StatefulSet) error {
+	klog.InfoS("Processing create notification", "Namespace", ss.Namespace, "StatefulSet", ss.Name)
+
+	ipPools := c.getIPPoolsForStatefulSet(ss)
+
+	if ipPools == nil {
+		// nothing to preallocate
+		return nil
+	}
+
+	if len(ipPools) > 1 {
+		return fmt.Errorf("annotation of multiple IP Pools is not supported")
+	}
+
+	// Only one pool is supported for now. Dual stack support coming in future.
+	ipPoolName := ipPools[0]
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPoolName, c.crdClient, c.ipPoolLister)
+	if err != nil {
+		return fmt.Errorf("failed to find IP Pool %s: %s", ipPoolName, err)
+	}
+
+	size := int(*ss.Spec.Replicas)
+	err = allocator.AllocateStatefulSet(ss.Namespace, ss.Name, size)
+	if err != nil {
+		return fmt.Errorf("failed to preallocate continuous IP space of size %d from Pool %s: %s", size, ipPoolName, err)
+	}
+
+	return nil
+}
 func (c *AntreaIPAMController) statefulSetWorker() {
 	for c.processNextStatefulSetWorkItem() {
 	}
 }
 
 func (c *AntreaIPAMController) processNextStatefulSetWorkItem() bool {
-	key, quit := c.statefulSetCleanupQueue.Get()
+	key, quit := c.statefulSetQueue.Get()
 	if quit {
 		return false
 	}
 
-	defer c.statefulSetCleanupQueue.Done(key)
-	err := c.cleanIPPoolForStatefulSet(key.(string))
+	defer c.statefulSetQueue.Done(key)
 
-	if err != nil {
-		// Put the item back on the workqueue to handle any transient errors.
-		c.statefulSetCleanupQueue.AddRateLimited(key)
-		klog.ErrorS(err, "Failed to clean IP Pool for StatefulSet", "StatefulSet", key)
-		return true
+	namespacedName := key.(string)[1:]
+
+	if key.(string)[:1] == delEventIndication {
+		err := c.cleanIPPoolForStatefulSet(namespacedName)
+		if err != nil {
+			// We can not put the event back to the queue because of
+			// potential recreate scenario (delete + create).
+			// We rely on garbage collector to clear IPs in case of
+			// error.
+			c.statefulSetQueue.Forget(key)
+			klog.ErrorS(err, "failed to clean IP Pool", "StatefulSet", key)
```

Comment:
```suggestion
			klog.ErrorS(err, "Failed to clean IP allocation of StatefulSet from IP Pool", "StatefulSet", key)
```

Reply:
Can you please clarify the convention regarding capitalizing error messages. @jianjuns suggests error message should start form lowercase https://github.com/antrea-io/antrea/pull/3281#discussion_r831501281. Does this depend on whether error message is printed to the log?

Reply:
All log message (no matter info, error, warn) should be capitalized. I meant message of an **Error** (e.g. fmt.Errorf()) should not.

Reply:
Got it, thanks